File: queue/Deque.java
public class Deque {
    private int rear, front, size, capacity;
    private int deque[];

    Deque(int capacity){
        this.capacity = capacity;
        this.rear = -1;
        this.front = -1;
        this.size = 0;
        this.deque = new int[capacity];
    }

    public Boolean isEmpty(){
        /** This method is used to check if the Deque is Empty */
        return this.front == -1 && this.rear == -1;
    }

    public Boolean isFull(){
        /** This method is used to check if the Deque is full or not */
        return (this.front == 0 && this.rear == this.capacity -1) || (this.front == this.rear + 1);
    }

    public void insertFront(int item){
        /** This method is used to add elements at the front of the deque*/
        if (this.isFull()){
            System.out.println("Overflow");
            System.exit(1);
        }
        if (this.isEmpty()){
            this.front = 0;
            this.rear = 0;
        }
        if (this.front == 0){
            this.front = this.capacity - 1;
        }else{
            this.front--;
        }
        this.deque[this.front] = item;
        this.size++;
        System.out.println("Inserted at front : "+item);
    }

    public void insertRear(int item){
        /** This method is used to add elements at the rear side of the deque */
        if (this.isFull()){
            System.out.println("Overflow");
            System.exit(1);
        }
        if (this.isEmpty()){
            this.front = 0;
            this.rear = 0;
        }
        else if (this.rear == this.capacity - 1){
            this.rear = 0;
        }else{
            this.rear++;
        }
        this.deque[this.rear] = item;
        this.size++;
        System.out.println("Inserted at rear : "+item);
    }

    public int deleteFront(){
        /** This method is used to delete elements at the front side of the deque */
        if (this.isEmpty()){
            System.out.println("Underflow");
            System.exit(1);
        }
        int temp = this.deque[this.front];
        if(this.front == this.rear){
            this.front = -1;
            this.rear = -1;
        }else {
            this.front = (this.front + 1) % this.capacity;
        }
        this.size--;
        System.out.println("Removed from front : "+temp);
        return temp;
    }

    public int deleteRear(){
        /** This method is used to delete elements at the rear side of the deque */
        if (this.isEmpty()){
            System.out.println("Underflow");
            System.exit(1);
        }
        int temp = this.deque[this.rear];
        if(this.front == this.rear){
            this.front = -1;
            this.rear = -1;
        }else if (this.rear == 0){
            this.rear = this.capacity -1;
        }else{
            this.rear--;
        }
        this.size--;
        System.out.println("Removed from rear : "+temp);
        return temp;
    }
    public int getFront(){
        /** This method returns the front element of the deque */
        if (this.isEmpty()){
            System.out.println("Underflow");
            System.exit(1);
        }
        return this.deque[this.front];
    }
    public int getRear(){
        /** This method returns the rear element of the deque */
        if (this.isEmpty()){
            System.out.println("Underflow");
            System.exit(1);
        }
        return this.deque[this.rear];
    }
    public int size(){
        /** This method returns the size of the deque */
        return this.size;
    }

    public void display() {
        /** This method is used to desplay the elements of the deque */
        if (this.isEmpty()) {
            System.out.println("Underflow");
            System.exit(1);
        }

        if (this.front > this.rear) {
            int i = this.front;
            System.out.print("Deque : " + this.deque[i]);
            do {
                i = (i + 1) % this.capacity;
                System.out.print(" ~ " + this.deque[i]);
            } while (i != this.rear);
            System.out.println("");

        }else{
            int i = this.front;
            System.out.print("Deque : " + this.deque[i]);
            while (i < this.rear) {
                i++;
                System.out.print(" ~ " + this.deque[i]);
            }
            System.out.println("");

        }

    }



}


class Test {
    public static void main(String args[]) {
        Deque deque = new Deque(5);
        //System.out.println(deque.getFront()); // will raise error Underflow
        //System.out.println(deque.getRear()); // will raise error Underflow
        //System.out.println("delete front "+ deque.deleteFront()); // will raise error Underflow
        //System.out.println("delete rear  "+ deque.deleteRear()); // will raise error Underflow

        deque.insertRear(5);
        deque.insertRear(10);
        deque.insertRear(15);
        System.out.println(deque.getFront());
        System.out.println(deque.getRear());
        System.out.println(deque.size());
        deque.display();
        deque.deleteFront();
        deque.deleteFront();
        System.out.println(deque.getFront());
        System.out.println(deque.getRear());
        System.out.println(deque.size());
        deque.display();

        deque.insertFront(20);
        deque.insertFront(25);
        deque.insertFront(30);
        deque.insertFront(35);
        System.out.println(deque.getFront());
        System.out.println(deque.getRear());
        System.out.println(deque.size());
        deque.display();
        //deque.insertFront(40); // will raise error Overflow
        //deque.insertRear(40);// will raise error Overflow
        deque.deleteFront();
        deque.deleteRear();

        System.out.println(deque.getFront());
        System.out.println(deque.getRear());
        System.out.println(deque.size());
        deque.display();
    }
}

File: queue/deque.py
class Deque:
    def __init__(self, capacity):
        """ This is the constructor to initialize the private attributes of the class """
        self.__front = -1
        self.__rear = -1
        self.__size = 0
        self.__deque = [None] * capacity
        self.__capacity = capacity
    
    def isFull(self):
        """This method checks if the deque is full or not"""
        return (self.__front == 0 and self.__rear == self.__capacity - 1) or (self.__front == self.__rear + 1)
    
    def isEmpty(self):
        """This method checks if the deque is empty or not"""
        return self.__front == -1 and self.__rear == -1
    
    def insertFront(self, item):
        """This method inserts elemnts at the front of the deque"""
        if self.isFull():
            print("Overflow")
        else:
            if self.isEmpty():
                self.__front = 0
                self.__rear = 0
                
            elif self.__front == 0:
                self.__front = self.__capacity - 1
            else:
                self.__front -= 1
            
            self.__deque[self.__front] = item
            self.__size += 1
            print("Inserted at front : {}".format(item))    
                    
    def insertRear(self, item):
        """This method inserts elemnts at the rear of the deque"""
        if self.isFull():
            print("Overflow")
        else:
            if self.isEmpty():
                self.__front = 0
                self.__rear = 0
            elif self.__rear == self.__capacity - 1:
                self.__rear = 0
            else:
                self.__rear += 1
            
            self.__deque[self.__rear] = item
            self.__size += 1
            print("Inserted at rear : {}".format(item))            

    def deleteFront(self):
        """This method removes elemnts from the front of the deque"""
        if self.isEmpty():
            print("Underflow")
        else:
            temp = self.__deque[self.__front]
            if self.__front == self.__rear:
                self.__front = -1
                self.__rear = -1
            elif self.__front == self.__capacity - 1:
                self.__front = 0
            else:
                self.__front += 1
                    
            self.__size -= 1
            print("Removed from front : {}".format(temp))    
            return temp        

    def deleteRear(self):
        """This method removes elemnts from the rear of the deque"""
        if self.isEmpty():
            print("Underflow")
        else:
            temp = self.__deque[self.__rear]
            if self.__front == self.__rear:
                self.__front = -1
                self.__rear = -1
            elif self.__rear == 0:
                self.__rear = self.__capacity - 1
            else:
                self.__rear -= 1
                    
            self.__size -= 1
            print("Removed from rear : {}".format(temp))    
            return temp        

    def getFront(self):
        """This method to return the front element of the deque"""
        if self.isEmpty():
            print("Underflow")
        else:
            return self.__deque[self.__front]

    def getRear(self):
        """This method to return the rear element of the deque"""
        if self.isEmpty():
            print("Underflow")
        else:
            return self.__deque[self.__rear]
        
    def size(self):
        """This method returns the size of the deque"""
        return self.__size
    
    def display(self):
        """This method displays the elements of the deque"""
        if self.isEmpty():
            print("Underflow")
        elif self.__front <= self.__rear:
            i = self.__front
            final_string = "Deque :"
            while(i <= self.__rear):
                final_string = "{} {},".format(final_string, self.__deque[i])
                i += 1
            print(final_string[:-1])
        else:
            i = self.__front
            final_string = "Deque : {}, ".format(self.__deque[i])
            while(i != self.__rear):
                i = (i + 1) % self.__capacity
                final_string = "{} {},".format(final_string, self.__deque[i])
            print(final_string[:-1])

    

if __name__ == '__main__':
    deque = Deque(5);
    print("front element : ",deque.getFront())
    print("rear element : ",deque.getRear())
    print("delete front ", deque.deleteFront())
    print("delete rear  ", deque.deleteRear())   
    print("size  ", deque.size())   
    deque.display()
    
    deque.insertRear(5);
    deque.insertRear(10);
    deque.insertRear(15);
    print("front element : ",deque.getFront())
    print("rear element : ",deque.getRear())
    print("size :  ", deque.size())   
    deque.display()
    
    deque.deleteFront();
    deque.deleteFront();
    print("front element : ",deque.getFront())
    print("rear element : ",deque.getRear())
    print("size :  ", deque.size())   
    deque.display()
    
    deque.insertFront(20);
    deque.insertFront(25);
    deque.insertFront(30);
    deque.insertFront(35);   
    print("front element : ",deque.getFront())
    print("rear element : ",deque.getRear())
    print("size :  ", deque.size())   
    deque.display()
    
    deque.deleteFront();
    deque.deleteRear();

    print("front element : ",deque.getFront())
    print("rear element : ",deque.getRear())
    print("size :  ", deque.size())   
    deque.display()
